Oxb 210/optimize workspace add

rpschoenburg opened the pull request:
<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Release Notes

* **New Features**
  * Workspace file uploads now support environment-variable-controlled concurrency and retry configurations for improved upload performance and reliability.
  * Enhanced batch upload error tracking enables per-file error reporting and targeted retry handling.

* **Improvements**
  * Added observability and tracing support for better performance monitoring and diagnostics.
  * Implemented concurrent processing for multipart batch uploads to improve upload throughput.
  * Expanded logging capabilities for improved debugging and error diagnostics during file operations.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

benartuso reviewed (approved):
love seeing the dtype icons in the cli, works for me!

rpschoenburg committed:
Refactor benchmarking module and add tests

Rebuilds the benchmark module to replace the criterion_group and
criterion_main macros with a custom main module that allows individual
benchmarks to be setup and ran without compiling the code for the other
ones. As well, the number of iterations and an optional data path can
now be specified via environment variables. Finally, benchmarks for push
and workspace_add were added

rpschoenburg committed:
Updates for remote workflow benchmarks

rpschoenburg committed:
Refactor workspace add to use MPSC pipeline

Rebuilds the core logic for adding small files to workspaces,
implementing an MPSC pipeline to maximize utilization of the client.
The producer builds multiparts, offloading the work to a
blocking thread pool, while the consumer uploads the multiparts and
then stages the nodes. Err files are collected from both, which are
logged.

rpschoenburg committed:
Batch upload spawns new tasks for every file

rpschoenburg committed:
Linters

rpschoenburg committed:
Clear files_to_stage on every iteration of workspace add

rpschoenburg committed:
Offload version files to blocking threadpool

Refactor save_multiparts in the versions controller to spawn a blocking
task for each received multipart, offloading these to a separate
threadpool. This improves the tokio scheduler's ability to schedule
tasks efficiently. Errors are still captured per-file, using a mutex to
guard the shared err_files Vec

rpschoenburg committed:
Implement backpressure and retry for workspace add

Uses a semaphore to limit the amount of concurrent uploads allowed in
the core logic of parallel_batched_small_file_upload. The amount
defaults to 30, but can be configured via an environment variable.

Refactors workspace add to collect files that fail at any point of the
workspace add pipeline and retry them at the end of the function.
Currently, a single file failing after this retry will result in the
operation returning an error

rpschoenburg committed:
Merge branch 'main' into WIP/workspace-add

rpschoenburg committed:
Add progress bar for workspace add

rpschoenburg committed:
Fixes for CodeRabbit

rpschoenburg committed:
Merge branch 'main' into OXB-210/optimize_workspace_add
